source/agora/common/BitField.d

```diff
@@ -0,0 +1,191 @@
+/*******************************************************************************
+
+    Contains a BitField implementation based on Ocean's BitArray struct.
+    The choice of Ocean's over Phobos' implementation of BitArray is
+    because the Phobos implementation requires the backing store length to be
+    a multiple of size_t, which makes it platform-specific and hard to code for.
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.BitField;
+
+import agora.common.Deserializer;
+import agora.common.Hash;
+import agora.common.Serializer;
+
+import ocean.core.BitArray;
+
+import std.algorithm;
+import std.math;
+
+/// Ditto
+struct BitField
+{
+    /// Bitfield implementation
+    public BitArray bit_array;
+
+    /// Expose the bitfield API
+    public alias bit_array this;
+
+    /// Backing store (BitArray requires integer arrays)
+    private uint[] storage;
+
+    /// The number of bits used (workaround for: BitArray.length() is not @nogc)
+    private size_t num_bits;
+
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            num_bits = the number of bits to be used.
+
+    ***************************************************************************/
+
+    public this ( size_t num_bits ) @trusted
+    {
+        // BitArray uses ints as storage, we have to use multiples of 32
+        this.storage.length = (num_bits + 31) / 32;
+        this.num_bits = num_bits;
+        this.bit_array.initialize(this.storage, num_bits);
+    }
+
+    /***************************************************************************
+
+        Serialization. The data is packed tightly by
+        removing unused bytes from the uint[] storage,
+        and the length of bits is encoded instead of
+        the length of bytes.
+
+        Params:
+            dg = serialize function accumulator
+
+    ***************************************************************************/
+
+    public void serialize (scope SerializeDg dg) const @trusted
+    {
+        serializePart(this.num_bits, dg);
+        serializePart(cast(ubyte[])this.storage, dg);
```

Comment:
This breaks on different endianness

Reply:
It's a shame we don't have a big endian system in the CI then.

Reply:
Are the existing serialization routines safe w.r.t. endianness?

Reply:
For example:

```D
dg((cast(ubyte*)&record)[0 .. long.sizeof]);
```

I think this isn't safe?

Reply:
> Are the existing serialization routines safe w.r.t. endianness?

Probably not.

> It's a shame we don't have a big endian system in the CI then.

I *think* we could set a MIPS one but it'll take and it has lots of dependency, so let's leave that fun for later.